Header X-Consul-Lastcontact integer overflow

The Java stracktrace is the following 

Caused by: java.lang.NumberFormatException: For input string: "18446744073709551613"
        at java.lang.NumberFormatException.forInputString(NumberFormatException.java:65) ~[na:1.8.0_45]
        at java.lang.Long.parseLong(Long.java:592) ~[na:1.8.0_45]
        at java.lang.Long.valueOf(Long.java:803) ~[na:1.8.0_45]
        at com.orbitz.consul.util.Http.consulResponse(Http.java:110) ~[tiko-core-2.52-20170522.155427-46-with-deps.jar:na]
        at com.orbitz.consul.util.Http.extractConsulResponse(Http.java:61) ~[XXX-2.52-20170522.155427-46-with-deps.jar:na]
        at com.orbitz.consul.HealthClient.getHealthyServiceInstances(HealthClient.java:317) ~[XXX-2.52-20170522.155427-46-with-deps.jar:na]
        at com.orbitz.consul.HealthClient.getHealthyServiceInstances(HealthClient.java:300) ~[XXX-2.52-20170522.155427-46-with-deps.jar:na]

The value in the header cannot be parsed to a standard java long type since it is exactly 2^64 -3.
This occurred spontaneously after running a cluster for 1 day. I have been using consul for a while (2-3 months) and this error never occurred before. 
Here's the log of Consul. The error occurred at 16:20, we can see a WARN statement.

 2017/05/23 16:17:08 [INFO] raft: Starting snapshot up to 86824435
    2017/05/23 16:17:08 [INFO] snapshot: Creating new snapshot at /data/consul/data_dir/raft/snapshots/110-86824435-1495549028957.tmp
    2017/05/23 16:17:09 [INFO] snapshot: reaping snapshot /data/consul/data_dir/raft/snapshots/110-86746753-1495548654476
    2017/05/23 16:17:09 [INFO] raft: Compacting logs from 86778436 to 86814205
    2017/05/23 16:17:09 [INFO] raft: Snapshot to 86824435 complete
    2017/05/23 16:20:22 [INFO] consul.fsm: snapshot created in 22.045<C2><B5>s
    2017/05/23 16:20:22 [INFO] raft: Starting snapshot up to 86861757
    2017/05/23 16:20:22 [INFO] snapshot: Creating new snapshot at /data/consul/data_dir/raft/snapshots/110-86861757-1495549222963.tmp
    2017/05/23 16:20:23 [INFO] snapshot: reaping snapshot /data/consul/data_dir/raft/snapshots/110-86788659-1495548856084
    2017/05/23 16:20:23 [INFO] raft: Compacting logs from 86814206 to 86851527
    2017/05/23 16:20:23 [INFO] raft: Snapshot to 86861757 complete
    2017/05/23 16:20:55 [WARN] agent: Check 'core-node-health' missed TTL, is now critical
    2017/05/23 16:20:55 [INFO] agent: Synced check 'core-node-health'
    2017/05/23 16:22:58 [INFO] consul.fsm: snapshot created in 24.637<C2><B5>s
    2017/05/23 16:22:58 [INFO] raft: Starting snapshot up to 86896041
    2017/05/23 16:22:58 [INFO] snapshot: Creating new snapshot at /data/consul/data_dir/raft/snapshots/110-86896041-1495549378723.tmp
    2017/05/23 16:22:58 [INFO] snapshot: reaping snapshot /data/consul/data_dir/raft/snapshots/110-86824435-1495549028957
    2017/05/23 16:22:58 [INFO] raft: Compacting logs from 86851528 to 86885811

--- a/command/agent/http.go
+++ b/command/agent/http.go
@@ -311,6 +311,9 @@ func setKnownLeader(resp http.ResponseWriter, known bool) {
 
 // setLastContact is used to set the last contact header
 func setLastContact(resp http.ResponseWriter, last time.Duration) {
+	if last < 0 {
+		last = 0
+	}
 	lastMsec := uint64(last / time.Millisecond)
 	resp.Header().Set("X-Consul-LastContact", strconv.FormatUint(lastMsec, 10))
 }

--- a/command/agent/http_test.go
+++ b/command/agent/http_test.go
@@ -150,11 +150,25 @@ func TestSetKnownLeader(t *testing.T) {
 
 func TestSetLastContact(t *testing.T) {
 	t.Parallel()
-	resp := httptest.NewRecorder()
-	setLastContact(resp, 123456*time.Microsecond)
-	header := resp.Header().Get("X-Consul-LastContact")
-	if header != "123" {
-		t.Fatalf("Bad: %v", header)
+	tests := []struct {
+		desc string
+		d    time.Duration
+		h    string
+	}{
+		{"neg", -1, "0"},
+		{"zero", 0, "0"},
+		{"pos", 123 * time.Millisecond, "123"},
+		{"pos ms only", 123456 * time.Microsecond, "123"},
+	}
+	for _, tt := range tests {
+		t.Run(tt.desc, func(t *testing.T) {
+			resp := httptest.NewRecorder()
+			setLastContact(resp, tt.d)
+			header := resp.Header().Get("X-Consul-LastContact")
+			if got, want := header, tt.h; got != want {
+				t.Fatalf("got X-Consul-LastContact header %q want %q", got, want)
+			}
+		})
 	}
 }
 

